### ISSUE TITLE
Preserve formatting when serializing discriminated unions without fields

### ISSUE THREAD

**Issue by erik-lundgren** [microsoft/fsharplu#58]
When serializing discriminated unions without any fields, the current behaviour is to use the Newtonsoft.Json `ContractResolver`'s `GetResolvedPropertyName` for serialization of the case name.

This means that when serializing using `CamelCasePropertyNamesContractResolver`
```
type FooBar = Foo | Bar

{
    Foo: Bar
}
```
Becomes:
```
{
    "foo": "bar"
}
```

I would prefer that the serialization instead produces this:
```
{
    "foo": "Bar"
}
```

As I believe it better conveys the value.

The code causing this is likely at or near [Compact.fs:151](https://github.com/Microsoft/fsharplu/blob/master/FSharpLu.Json/Compact.fs#L151).

**Comment by erik-lundgren** [comment]
I realize this is the opposite of #49 and while the behaviour is consistent within the package (DUs with fields get converted to object keys using the property name camel case formatting) it does not seem like a consistent behaviour with Newtonsoft.Json, which will change json property casing but preserve json value casing.

Shortening (from Newtonsoft.Json)
```
"someValue": {
    "case": "Foo"
}
```

to 

```
"someValue": "foo"
```

was very unexpected for me

**Comment by blumu** [comment]
@erik-lundgren I don't quite follow the example in the bug description. The example given is not a proper F# value:
```fs
{
    Foo: Bar
}
```
Could you maybe clarify by given the full repro with an F# code example?

**Comment by blumu** [comment]
@erik-lundgren  This is what the FSharpLu.Compact serializer currently produces for field-less discriminated unions:
```fs
type FooBar = Foo | Bar

Compact.serialize Foo
Compact.serialize Bar
>
type FooBar =
  | Foo
  | Bar
> 
val it : string = ""Foo""

val it : string = ""Bar""
```
I am guessing you were thinking of a more involved data-type?

**Comment by Nexith** [comment]
This is using the Strict serialize yet not avaiable in the NuGet package added by PR #54, that is using the `CamelCasePropertyNamesContractResolver`.
```fsharp
type Foo =
    | Bar

type Foobar = { 
    Foo: Foo
    Bar: string
}
let foo = {
    Foo = Foo.Bar
    Bar = "Bar"
}
Strict.serialize(foo)
```

Produces the following JSON, where "foo" should have the value "Bar" and not "bar". The contract resolver is touching the value of the property which it should not do. 
```json
{
    "foo": "bar",
    "bar": "Bar"
}
```

**Comment by blumu** [comment]
@Nexith  Thanks for sharing the repro. This makes sense. What would then be the expected serialization for DU with both field-less and field cases? For example take type
```fs
type Foo =
| Raymond of string
| Bar
```
Should `Raymond "bar"` be serialized with camel casing, and `Bar` be serialized with normal casing?
The current implementation gives:
```fs
[ Raymond "Bar"; Foo.Bar ] |> Compact.Strict.serialize
> 
val it : string = "[
  {
    "raymond": "Bar"
  },
  "bar"
]"
```
where both cases of the discriminated unions are Caml-cased.

**Comment by erik-lundgren** [comment]
My reasoning is 
1. The serialization should first abide by serialization configuration
1. The serialization should then preserve as much meaning as possible.
1. The json properties are subject to formatting by the property name resolver but json values are not.
1. Therefore the serialization should preserve meaning for json values but abide by serialization configuration for property name formatting.

Given that reasoning, when the `CamelCasePropertyNameResolver` is used, then my opinion would be that the serialization example in your post makes the most sense when instead producing:

```
[ Raymond "Bar"; Foo.Bar ] |> Compact.Strict.serialize
> 
val it : string = "[
  {
    "raymond": "Bar"
  },
  "Bar"
]"
```
where configuration has been applied for `Raymond` and meaning has been preserved for `Bar`.

**Comment by johlrich** [comment]
I agree with points 1 and 2 above, but I think of the what's being preserved as the original DU's semantics of a case with fields and not the semantics of the json encoding. With that viewpoint, I believe the consistency of the case is important regardless of the position in the encoding.

One idea is changing out the settings for a different union encoder that captures the case in a consistent location like a `{case, field}` object and not use the CompactUnionJsonConverter. 

I thought about switching to that encoding at one point during development to gain consistency across types instead of string vs object, but instead settled on creating a decodeUnion helper that translates the compact union encoding into a `{case, field}` object. For example, the `{case, field}` objects are way more convenient to use with something like TypeScript's union while retaining the benefits of the compact encoding across the backend. 

Another idea could be having the CompactUnionJsonConverter with an additional option to override the setting so folks can reuse the rest of the work in there with custom settings?

**Comment by erik-lundgren** [comment]
@johlrich Interesting solution, I understand why it makes a lot of sense for TypeScript. I will be implementing a similar solution in my project, and I don't see this issue being a blocker for me. I just think it looks better and I expected the serializer to preserve casing for my DUs.

Regarding what's being preserved I don't understand your argument fully. Currently the DU is not preserved in either json properties nor values as case is changed in both. I agree that the case always changing is some form of consistency, but it seems very internal to this project and is non-obvious to a new user. Again, consider the following JSON serialization example taken from the [wiki](https://github.com/Microsoft/fsharplu/wiki/fsharplu.json):

| Value | Default (Json.net) | Compact
| ----- | ------------------ | ----------
| `Foo` |  `{  "Case": "Foo" }` |  `"Foo"`| 

I'm arguing that the following transformation is not obvious or expected for a user of this library when using the `CamelCasePropertyNamesContractResolver`:

| Value | Default (Json.net) | Compact
| ----- | ------------------ | ----------
| `Foo` |  `{  "case": "Foo" }` |  `"foo"`| 

I believe preserving case here is important and to me this is the expected transformation:

| Value | Default (Json.net) | Compact
| ----- | ------------------ | ----------
| `Foo` |  `{  "case": "Foo" }` |  `"Foo"`| 

I also believe, as outlined above, that it is important to follow user configuration. If the user wants camel cased property names, the serialization should respect that over preserving union casing for properties. My expectation would be the following when using the `CamelCasePropertyNamesContractResolver` (adapted from my original post in this issue):

Value (F# record with one DU field):
```
{
    Foo: Bar
}
```

Default (Json.net)
```
{
    "foo": { 
        "case": "Bar"
     }
}
```

Compact
```
{
    "foo": "bar"
}
```

I don't believe it is obvious to a user of this library that the casing of the *value* changes when using a *property* name contract resolver.

**Comment by blumu** [comment]
@erik-lundgren I agree that _property_ name contract resolver should only affect properties and not values. So as you suggested, if `CamelCasePropertyNamesContractResolver` is used then we should go with the following serialization for `[ Raymond "Bar"; Foo.Bar ]`:
```
val it : string = "[
  {
    "raymond": "Bar"
  },
  "Bar"
]"
```
Before closing PR #60 we want to add unit tests to cover the casing behaviour expected in this example.

Eventually we should add a configurable option to determine how casing is handled for values, to cover for instance the situation described by @johlrich. Perhaps we can do that in a separate PR.

**Comment by johlrich** [comment]
Happy to help land additional options to the converter to allow us to keep case names being the same if it doesn't make it with this PR. What do you think they would be called?

**Comment by blumu** [comment]
Thanks for the offer @johlrich I've suggested the name `camelCaseValues` in comments of PR #60:
```
type CompactUnionJsonConverter(?tupleAsHeterogeneousArray:bool, ?camelCaseValues:bool) =
      let tupleAsHeterogeneousArray:bool= defaultArg tupleAsHeterogeneousArray false
      ...
     | [||] when tupleAsHeterogeneousArray  -> writer.WriteValue(case.Name)
     | [||] when not tupleAsHeterogeneousArray  -> writer.WriteValue(convertName case.Name)
      ...
```

**Comment by erik-lundgren** [comment]
@blumu I've implemented the suggestions in PR #60
